app/clients/zendesk_sell.py

```diff
@@ -82,10 +110,113 @@ def send_lead(self, contact: ContactRequest) -> int:
             return response.status_code
         except requests.RequestException as e:
             content = json.loads(response.content)
-            self.app.logger.warning(f"Failed to create zendesk sell lead: {content['errors']}")
+            current_app.logger.warning(f"Failed to create zendesk sell lead: {content['errors']}")
             raise e
-        except NotImplementedError:
-            # There are cases in development when we do not want to send to zendesk sell lead creation
-            # because configuration is not defined, lets return a 200 OK
-            self.app.logger.warning('Did not send lead to zendesk')
-            return 200
+
+    def upsert_contact(self, user: User) -> (Optional[int], bool):
+
+        # The API and field definitions are defined here: https://developers.getbase.com/docs/rest/reference/contacts
+        if not self.api_url or not self.token:
+            current_app.logger.warning('Did not upsert contact to zendesk')
+            return None, False
+
+        try:
+            response = requests.post(
+                url=urljoin(self.api_url, f'/v2/contacts/upsert?email={user.email_address}'),
+                data=json.dumps(ZenDeskSell._generate_contact_data(user)),
+                headers={'Accept': 'application/json', 'Content-Type': 'application/json'},
+                auth=BearerAuth(token=self.token),
+                timeout=5
```

Comment:
WDYT about refactoring these HTTP calls a bit? Headers, auth, timeout, checking if the API URL is set etc are duplicated quite a few times now. You could have a private method accepting the HTTP method, URI and data maybe? Could look at https://github.com/alphagov/notifications-python-client/blob/master/notifications_python_client/base.py for some inspiration.

Reply:
yes absolutely, DRY.